### PR TITLE
New Parser: Add restricted and dictionary types

### PR DESCRIPTION
**Pull request by turbolent**
Work towards dapperlabs/flow-go#2193

### CODE REVIEW

File: runtime/parser2/declaration.go [hunk 1]
```diff
@@ -157,32 +164,60 @@ func parseParameterList(p *parser) (parameterList *ast.ParameterList) {
...
 	p.next()
 
 	var endPos ast.Position
 
+	expectParameter := true
+
 	atEnd := false
 	for !atEnd {
 		p.skipSpaceAndComments(true)
 		switch p.current.Type {
```
Comment (benjaminkvm) [inline runtime/parser2/declaration.go:184]: We will also need to implement `lexer.TokenUnderscore` to handle un-labeled parameters.

Comment (benjaminkvm) [reply runtime/parser2/declaration.go:184]: Handling which will likely live in `parseParameter` I assume?

Comment (turbolent) [reply runtime/parser2/declaration.go:184]: We don't actually need to differentiate between underscore `_` and other identifiers on the lexer level. Identifiers may start with an underscore and may only contain one character, so it already covers the argument label case. See https://github.com/onflow/cadence/blob/505dc9b920954018e75410c4d26c0218bf9e0bc4/runtime/parser2/lexer/state.go#L91-L92

Comment (benjaminkvm) [reply runtime/parser2/declaration.go:184]: Cool wasn't sure if `_` was considered an identifier -- looks like it is 👍

**Comment by turbolent** [comment]
Yeah, this is seems to be so far the worst case in the parser in terms of ambiguity. I'll need to do a pass and document / comment at the end of the MVP stage, probably just copying the BNF like syntax we already have from the ANTLR grammar

**Pull request merged**